### PR TITLE
feat(sdk-core): commit to partial public key

### PR DESCRIPTION
**Pull request by zhongxishen**
TICKET: HSM-82

### CODE REVIEW

**Review by alebusse — COMMENTED** [review]
Could you please fix the merge conflict?

**Comment by zhongxishen** [comment]
> Could you please fix the merge conflict?

Done

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts [hunk 1]
```diff
@@ -145,6 +147,102 @@ export default class Ecdsa {
...
+   * @returns {UYShare} Returns share of private u value and commitment to public y value.
+   */
+  keyShareStep1(index: number, threshold: number, numShares: number, seed?: Buffer): UYShare {
+    if (!(index > 0 && index <= numShares && threshold <= numShares && threshold === 2)) {
+      throw 'Invalid KeyShare Config';
+    }
+
+    if (seed && seed.length < 64) {
+      throw new Error('Seed must have a length of at least 64 bytes');
+    }
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:165]: Nit: should have tests for these non-happy paths (throwing the errors)

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts [hunk 2]
```diff
@@ -145,6 +147,102 @@ export default class Ecdsa {
...
+
+    // Accept a 64 byte seed and create an extended private key from that seed
+    const secretKey = seedWithValidLength && bip32.fromSeed(seedWithValidLength);
+    const u =
+      (secretKey && secretKey.privateKey && bigIntFromU8ABE(new Uint8Array(secretKey.privateKey))) ??
+      Ecdsa.curve.scalarRandom();
+    const y = Ecdsa.curve.basePointMult(u);
+
+    const chaincode = (secretKey && secretKey.chainCode) ?? randomBytes(32);
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:180]:
```suggestion
    const chaincode = secretKey?.chainCode ?? randomBytes(32);
```

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts [hunk 3]
```diff
@@ -145,6 +147,102 @@ export default class Ecdsa {
...
+        keyShare.nShares[participantIndex] = {
+          i: participantIndex,
+          j: currentParticipant['i'],
+          n: publicKey.n.toString(16),
+          y: uyShare.comDecomY.decommitment.secret.toString('hex'),
+          decomY: uyShare.comDecomY.decommitment,
+          v: bigIntToBufferBE(v[0], 33).toString('hex'),
+          u: bigIntToBufferBE(uShares[participantIndex], 32).toString('hex'),
+          chaincode: uyShare.chaincode,
+        } as NShare;
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:240]: Why do we need to cast here?

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts [hunk 4]
```diff
@@ -153,6 +251,13 @@ export default class Ecdsa {
    * and y-shares to be used when generating signing shares.
    */
   keyCombine(pShare: PShare, nShares: NShare[]): KeyCombined {
+    nShares.map((share) => {
+      if (share.decomY && share.comY) {
+        if (!HashCommitment.verifyCommitment(share.comY, share.decomY)) {
+          throw new Error(`Could not verify commitment to y from participant ${share.j}`);
+        }
+      }
+    });
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:260]: Can we test this error being thrown?

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/types.ts [hunk 5]
```diff
@@ -57,6 +69,8 @@ export type NShare = SignIndex & {
   n: string;
   u: string; // shamir share of secret at j'th index
   y: string;
+  comY?: Buffer;
+  decomY?: HashDecommitment;
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/account-lib/mpc/tss/ecdsa/types.ts:73]: Nit: would be nice to have a comment on these fields, to know what they are/represent.

File: modules/sdk-core/test/unit/account-lib/mpc/tss/ecdsa/ecdsa.ts [hunk 6]
```diff
@@ -30,12 +30,42 @@ describe('ecdsa tss', function () {
...
+      ecdsa.keyShareStep1(1, 2, 3),
+      ecdsa.keyShareStep1(2, 2, 3),
+      ecdsa.keyShareStep1(3, 2, 3),
+    ];
+
+    // Exchange commitments of y_i (uyShare1.comDecomY.commitment, uyShare2.comDecomY.commitment, uyShare3.comDecomY.commitment).
+    // Make sure each party getting commitments from all other parties before proceeding to next step.
+    const yCommitments1 = {};
+    yCommitments1[2] = uyShare2.comDecomY.commitment;
+    yCommitments1[3] = uyShare3.comDecomY.commitment;
```
Comment (zahin-mohammad) [inline modules/sdk-core/test/unit/account-lib/mpc/tss/ecdsa/ecdsa.ts:43]: Nit:

```suggestion
    const yCommitments1 = {
    	2: uyShare2.comDecomY.commitment.
    	3: uyShare3.comDecomY.commitment,
    };
```

Same comment below.

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts (hunk 1 above)
Comment (zhongxishen) [reply modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts]: Added

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts (hunk 2 above)
Comment (zhongxishen) [reply modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts]: Done

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts (hunk 3 above)
Comment (zhongxishen) [reply modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts]: It was copied over from the original keyShare(), both removed.

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts (hunk 4 above)
Comment (zhongxishen) [reply modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:256]: Done

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/types.ts (hunk 5 above)
Comment (zhongxishen) [reply modules/sdk-core/src/account-lib/mpc/tss/ecdsa/types.ts]: Done

File: modules/sdk-core/test/unit/account-lib/mpc/tss/ecdsa/ecdsa.ts (hunk 6 above)
Comment (zhongxishen) [reply modules/sdk-core/test/unit/account-lib/mpc/tss/ecdsa/ecdsa.ts]: Done

**Comment by zhongxishen** [comment]
Check failed due to critical vulnerability report https://github.com/advisories/GHSA-67hx-6x53-jw92

Assuming it'll be fixed quickly on master, will rebase after.

**Comment by zhongxishen** [comment]
@zahin-mohammad rebased and ready to be reviewed again

**Comment by zahin-mohammad** [comment]
> @zahin-mohammad rebased and ready to be reviewed again

@zhongxishen do we still need this PR or can we close this out now?

**Review by zahin-mohammad — COMMENTED** [review]
(PR is in draft; flushing from review q)

**Review by alebusse — COMMENTED** [review]
+1

**Comment by zhongxishen** [comment]
Closing it since GG18 related work has stopped. 

Although technically it's a vulnerability of the key generation phase, so it may re-emerge if we switch to different signing protocol.

**Pull request closed without merging**